### PR TITLE
o/snapstate: send refresh hold data in store requests

### PR DESCRIPTION
**Pull request by ZeyadYasser**
The work was already implemented by @MiguelPires but was not merged because the store wasn't accepting the new metrics data yet. It is supported now in version 54 https://api.snapcraft.io/docs/.

I added a retry mechanism to support proxies that might be running older versions of the store API. This approach retries bad requests, with the new knowledge of the store version parsed from the response header `Snap-Store-Version` and removes the `held` field if store version is less than v54. This approach is very similar to was is done [here](https://github.com/snapcore/snapd/blob/f1e72fa0bd5eda1bb45161aec9b6edbfe30a9af4/store/store.go#L1124).
This approach is generic and can be used for other new fields using `minStoreVersion(storeVer, <MIN_VER>)`.

Variable names and comment formatting could use some refining, I am open to suggestions.

### CODE REVIEW

**Comment by ZeyadYasser** [comment]
Spread tests are failing due to a bug in the snapcraft API. I will update the PR when it is fixed.

File: store/store_action.go [hunk 1]
```diff
@@ -566,6 +571,17 @@ func (s *Store) snapAction(ctx context.Context, currentSnaps []*CurrentSnap, act
...
 
 	if resp.StatusCode != 200 {
+		// some fields might not be supported on proxies with old versions.
+		// we should retry with the snap store version known as we can now
+		// get it from the response header.
+		if resp.StatusCode == 400 && storeVer == nil {
+			verstr := resp.Header.Get("Snap-Store-Version")
+			ver, err := strconv.Atoi(verstr)
+			if err != nil {
+				logger.Debugf("Bogus Snap-Store-Version header %q.", verstr)
```
Comment (MiguelPires) [inline store/store_action.go:581]: When there's an error we usually start our error messages with "cannot". It's not super important here because it's not a user facing error but good to be aware of that pattern. I'd rephrase this because the value is more unexpected or wrong that bogus, something like: `cannot parse header value of Snap-Store-Version: expected int got %q`
Just a suggestion, doesn't have to be specifically that

File: store/store_action.go [hunk 2]
```diff
@@ -309,7 +314,7 @@ type AssertionResult struct {
 	StreamURLs []string
 }
 
-func (s *Store) snapAction(ctx context.Context, currentSnaps []*CurrentSnap, actions []*SnapAction, assertQuery AssertionQuery, toResolve map[asserts.Grouping][]*asserts.AtRevision, toResolveSeq map[asserts.Grouping][]*asserts.AtSequence, user *auth.UserState, opts *RefreshOptions) ([]SnapActionResult, []AssertionResult, error) {
+func (s *Store) snapAction(ctx context.Context, currentSnaps []*CurrentSnap, actions []*SnapAction, assertQuery AssertionQuery, toResolve map[asserts.Grouping][]*asserts.AtRevision, toResolveSeq map[asserts.Grouping][]*asserts.AtSequence, user *auth.UserState, opts *RefreshOptions, storeVer *int) ([]SnapActionResult, []AssertionResult, error) {
```
Comment (MiguelPires) [inline store/store_action.go:317]: Taking a pointer makes the zero value more obvious but I think an `int` would've been fine. We could ignore if it was zero (since no proxy will likely return it) or negative and take it into account otherwise. Just a opinion, maybe @pedronis disagrees

File: store/store_action.go [hunk 3]
```diff
@@ -347,6 +358,10 @@ func (s *Store) snapAction(ctx context.Context, currentSnaps []*CurrentSnap, act
 			CohortKey:        curSnap.CohortKey,
 			ValidationSets:   valsetKeys,
 		}
+		// `held` field was introduced in version 54 https://api.snapcraft.io/docs/
+		if len(curSnap.HeldBy) > 0 && minStoreVersion(storeVer, 54) {
```
Comment (MiguelPires) [inline store/store_action.go:362]: Then this would be explicit about defaulting to using the held-by when no version if specified

```suggestion
		if len(curSnap.HeldBy) > 0 && (storeVer <= 0 || storeVer >= 54) {
```

File: store/store_action.go [hunk 4]
```diff
@@ -289,6 +296,10 @@ func genInstanceKey(curSnap *CurrentSnap, salt string) (string, error) {
 	return fmt.Sprintf("%s:%s", curSnap.SnapID, enc), nil
 }
 
+func minStoreVersion(storeVer *int, minVer int) bool {
```
Comment (MiguelPires) [inline store/store_action.go:299]: If we end up keeping this, calling this `meetsVersion` (or something like that) would be more self-explanatory IMO

File: overlord/snapstate/storehelpers_test.go [hunk 5]
```diff
@@ -483,3 +485,74 @@ func (s *snapmgrTestSuite) TestInstallSizeRemotePrereq(c *C) {
...
+	mockInstalledSnap(c, st, snapCyaml, false)
+
+	//mockLastRefreshed(c, st, "2021-05-09T10:00:00Z", "snap-a", "snap-b", "snap-c")
+
+	//now, err := time.Parse(time.RFC3339, "2021-05-10T10:00:00Z")
+	//c.Assert(err, IsNil)
+	//restore := snapstate.MockTimeNow(func() time.Time {
+	//	return now
+	//})
+	//defer restore()
```
Comment (MiguelPires) [inline overlord/snapstate/storehelpers_test.go:505]: I don't remember exactly but I think this might be a left over from when HoldRefresh worked differently? I think it can be removed.

File: store/store_action.go (hunk 2 above)
Comment (ZeyadYasser) [reply store/store_action.go:317]: I am thinking that could allow malicious servers (mitm) to use this header to force snapd into an infinite recursion, and eventually crashing. for example always setting it to 0, or not setting the header at all. I actually noticed this when some tests crashed due not setting the snap version header as it was not relevant to their test case.
That was my reasoning when opting for passing a pointer.

File: store/store_action.go (hunk 3 above)
Comment (ZeyadYasser) [reply store/store_action.go:362]: Yes, I agree. Even if we end up using a pointer this would be more explicit `(storeVer == nil || *storeVer >= 54)`. Thank you!

File: store/store_action.go (hunk 4 above)
Comment (ZeyadYasser) [reply store/store_action.go]: I removed it completely, I totally agree on making the store version check explicit (e.g. `(storeVer == nil || *storeVer >= 54)`).

File: store/store_action.go (hunk 1 above)
Comment (ZeyadYasser) [reply store/store_action.go]: Updated, Thank you!

File: store/store_action.go (hunk 2 above)
Comment (MiguelPires) [reply store/store_action.go:313]: IMO we should have an explicit limit to retries. Right now, retries are limited because `strconv.Atoi` returns a zero on error (which we would get if the header was empty), which we take a reference of and then check against nil. That's not very clear IMO and my (personal) preference would be to be very explicit about it and check that the header carries a version that's worth retrying with. But we can wait and see if other reviewers feel differently. 
Also if we consider malicious server, we might have many other problems so I'm not sure that's something we usually include in our fault model.

Comment (ZeyadYasser) [reply store/store_action.go:313]: > my (personal) preference would be to be very explicit about it and check that the header carries a version that's worth retrying with.

Aha, I like this approach, thank you for explaining.

File: store/store_action.go [hunk 6]
```diff
@@ -556,6 +567,17 @@ func (s *Store) snapAction(ctx context.Context, currentSnaps []*CurrentSnap, act
...
 
 	if resp.StatusCode != 200 {
+		// some fields might not be supported on proxies with old versions.
+		// we should retry with the snap store version known as we can now
+		// get it from the response header.
+		if resp.StatusCode == 400 && storeVer == nil {
+			verstr := resp.Header.Get("Snap-Store-Version")
+			ver, err := strconv.Atoi(verstr)
+			if err != nil {
+				logger.Debugf("cannot parse header value of Snap-Store-Version: expected int got %q", verstr)
```
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply store/store_action.go]: Update, Thanks!

File: store/store_action_test.go [hunk 7]
```diff
@@ -2509,6 +2509,199 @@ func (s *storeActionSuite) TestSnapActionRefreshStableInstanceKey(c *C) {
...
+	}, nil, nil, &store.RefreshOptions{PrivacyKey: "123"})
+
+	c.Assert(err, IsNil)
+	c.Assert(results, HasLen, 1)
+	c.Assert(results[0].SnapName(), Equals, "hello-world")
+	c.Assert(results[0].InstanceName(), Equals, "hello-world")
+	c.Assert(results[0].Revision, Equals, snap.R(26))
+}
+
+func (s *storeActionSuite) TestSnapActionRefreshWithHeldUnupportedProxy(c *C) {
```
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply store/store_action_test.go]: Updated, Thanks!

File: store/store_action.go [hunk 8]
```diff
@@ -556,6 +567,18 @@ func (s *Store) snapAction(ctx context.Context, currentSnaps []*CurrentSnap, act
 	}
 
 	if resp.StatusCode != 200 {
+		// some fields might not be supported on proxies with old versions.
+		// we should retry with the snap store version known as we can now
+		// get it from the response header.
+		if resp.StatusCode == 400 && storeVer <= 0 {
```
Comment (pedronis) [inline store/store_action.go:573]: what about the error we were getting even with 54? was it a 400 or a 500? do we need to fallback to turn 54 into 53 perhaps if we reach here with that error combination? the question I suppose is whether there is a stable proxy version shipped with the problem or not yet?

Comment (pedronis) [reply store/store_action.go:573]: a different approach would to ask the store to bump the version (to 55) once all the issues are solved and use that here instead

File: overlord/snapstate/storehelpers.go [hunk 9]
```diff
@@ -781,6 +800,46 @@ func refreshCandidates(ctx context.Context, st *state.State, names []string, rev
...
+	if allSnapsHold != "" {
+		if allSnapsHold == "forever" {
+			allSnapsHoldTime = timeNow().Add(maxDuration)
+		} else {
+			allSnapsHoldTime, err = time.Parse(time.RFC3339, allSnapsHold)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
```
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:823]: do we have code like this somewhere in the autorefresh code, that perhaps we should turn into a helper?

File: overlord/snapstate/storehelpers.go [hunk 10]
```diff
@@ -781,6 +800,46 @@ func refreshCandidates(ctx context.Context, st *state.State, names []string, rev
 	return updates, stateByInstanceName, ignoreValidationByInstanceName, nil
 }
 
+// SnapHolds returns a map of held snaps to lists of holding snaps (including
+// "system" for user holds).
+func SnapHolds(st *state.State, snaps []string) (map[string][]string, error) {
```
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:805]: right now both HeldSnaps and SnapHolds are exported, neither is so far used outside snapstate atm though. Does it make sense to keep exporting both?

**Review by pedronis — COMMENTED** [review]
thanks, looking good, some questions though

File: overlord/snapstate/storehelpers.go (hunk 9 above)
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:823]: do we have code like this somewhere in the autorefresh code, that perhaps we should turn into a helper?

Comment (pedronis) [reply overlord/snapstate/storehelpers.go:823]: I selected the wrong code range, see the remade comment below

File: store/store_action.go (hunk 8 above)
Comment (ZeyadYasser) [reply store/store_action.go:573]: The solution was intended to be generic so other features might use in the future for different api versions, I can't think of a way that can be used to determine which API version to fallback to in case of failure with api known.
I like the deterministic nature of the version bump approach.

File: overlord/snapstate/storehelpers.go (hunk 10 above)
Comment (ZeyadYasser) [reply overlord/snapstate/storehelpers.go:805]: I didn't do the initial implementation maybe @MiguelPires has a more convincing reason, but I think the export was done to [facilitate testing](https://github.com/ZeyadYasser/snapd/commit/6e09eb12b61288eba5ab06faeed9fa89bd3cfe27#diff-8e164732575e20f0c572a4ff42ad8c71e885a6ecf24356bb31a27e5e6126d4e4R510) as testing package is `snapstate_test` instead of `snapstate`.

File: overlord/snapstate/storehelpers.go (hunk 9 above)
Comment (ZeyadYasser) [reply overlord/snapstate/storehelpers.go]: thanks, updated!

File: overlord/snapstate/storehelpers.go [hunk 11]
```diff
@@ -800,28 +800,37 @@ func refreshCandidates(ctx context.Context, st *state.State, names []string, rev
 	return updates, stateByInstanceName, ignoreValidationByInstanceName, nil
 }
 
-// SnapHolds returns a map of held snaps to lists of holding snaps (including
-// "system" for user holds).
-func SnapHolds(st *state.State, snaps []string) (map[string][]string, error) {
-	var allSnapsHold string
+func effectiveRefreshHold(st *state.State) (time.Time, error) {
```
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:803]: thanks for introducing this but autorefresh.go is probably a better placement for this, if you introduce it just after EffectiveRefreshHold

**Review by pedronis — COMMENTED** [review]
thanks, code placement comment

File: overlord/snapstate/storehelpers.go (hunk 10 above)
Comment (pedronis) [reply overlord/snapstate/storehelpers.go]: we can solve this in a follow-up when @MiguelPires is back to discuss, it's not a blocker as such

**Review by pedronis — APPROVED** [review]
thank you

**Comment by pedronis** [comment]
marking blocked until we are sure of the final API version number to use

File: store/store_action.go [hunk 12]
```diff
@@ -347,6 +354,10 @@ func (s *Store) snapAction(ctx context.Context, currentSnaps []*CurrentSnap, act
 			CohortKey:        curSnap.CohortKey,
 			ValidationSets:   valsetKeys,
 		}
+		// `held` field was introduced in version 54 https://api.snapcraft.io/docs/
+		if len(curSnap.HeldBy) > 0 && (storeVer <= 0 || storeVer >= 54) {
```
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply store/store_action.go]: issue is not fixed and api version was bumped to v55

**Comment by ZeyadYasser** [comment]
Issues in the store api were resolved and version was bumped to v55, removing blocked label.

**Review by MiguelPires — APPROVED** [review]
lgtm, thanks. Can you squash the last 6 commits into one before the merge please?

**Comment by ZeyadYasser** [comment]
> lgtm, thanks. Can you squash the last 6 commits into one before the merge please?

Thank you for the review, commits are now squashed
